src/GraphQLResponse.cpp

```diff
@@ -3,159 +3,200 @@
+			_data = { MapData {} };
 			break;
 
 		case Type::List:
-			_data = std::make_unique<TypedData>(TypedData { std::make_optional<ListData>() });
+			_data = { ListType {} };
 			break;
 
 		case Type::String:
-		case Type::EnumValue:
-			_data =
-				std::make_unique<TypedData>(TypedData { std::make_optional<StringOrEnumData>() });
+			_data = { StringData {} };
 			break;
 
-		case Type::Scalar:
-			_data = std::make_unique<TypedData>(TypedData { std::make_optional<ScalarData>() });
+		case Type::Null:
+			_data = { NullData {} };
 			break;
 
 		case Type::Boolean:
-			_data = std::make_unique<TypedData>(TypedData { BooleanType { false } });
+			_data = { BooleanType { false } };
 			break;
 
 		case Type::Int:
-			_data = std::make_unique<TypedData>(TypedData { IntType { 0 } });
+			_data = { IntType { 0 } };
 			break;
 
 		case Type::Float:
-			_data = std::make_unique<TypedData>(TypedData { FloatType { 0.0 } });
+			_data = { FloatType { 0.0 } };
 			break;
 
-		default:
+		case Type::EnumValue:
+			_data = { EnumData {} };
+			break;
+
+		case Type::Scalar:
+			_data = { ScalarData {} };
 			break;
 	}
 }
 
 Value::~Value()
 {
 	// The default destructor gets inlined and may use a different allocator to free Value's member
-	// variables than the graphqlservice module used to allocate them. So even though this could be
-	// omitted, declare it explicitly and define it in graphqlservice.
+	// variables than the graphqlresponse module used to allocate them. So even though this could be
+	// omitted, declare it explicitly and define it in graphqlresponse.
 }
 
 Value::Value(const char* value)
-	: _type(Type::String)
-	, _data(std::make_unique<TypedData>(
-		  TypedData { StringOrEnumData { StringType { value }, false } }))
+	: _data(TypeData { StringData { StringType { value }, false } })
 {
 }
 
 Value::Value(StringType&& value)
-	: _type(Type::String)
-	, _data(std::make_unique<TypedData>(TypedData { StringOrEnumData { std::move(value), false } }))
+	: _data(TypeData { StringData { std::move(value), false } })
 {
 }
 
 Value::Value(BooleanType value)
-	: _type(Type::Boolean)
-	, _data(std::make_unique<TypedData>(TypedData { value }))
+	: _data(TypeData { value })
 {
 }
 
 Value::Value(IntType value)
-	: _type(Type::Int)
-	, _data(std::make_unique<TypedData>(TypedData { value }))
+	: _data(TypeData { value })
 {
 }
 
 Value::Value(FloatType value)
-	: _type(Type::Float)
-	, _data(std::make_unique<TypedData>(TypedData { value }))
+	: _data(TypeData { value })
 {
 }
 
 Value::Value(Value&& other) noexcept
-	: _type(other.type())
-	, _data(std::move(other._data))
+	: _data(std::move(other._data))
 {
 }
 
 Value::Value(const Value& other)
-	: _type(other.type())
-	, _data(std::make_unique<TypedData>(other._data ? *other._data : TypedData {}))
 {
+	switch (other.type())
+	{
+		case Type::Map:
+		{
+			MapData copy {};
+
+			copy.map.reserve(other.size());
+			for (const auto& entry : other)
```

Comment:
did you test if this is triggering and how frequently?

In my PR I did some tests and found out that it wasn't needed if we Copy-on-Write, that's why I had a shared_ptr<> for the internal data, then copy became just a reference. During modification I did a `unique()` and if false I'd copy there, on demand (search for `willModify`).

After I did that, it turns that no copies were being done.

Reply:
I suspect that switching to a result type eliminated all of the use cases for copying. Maybe it can be a move-only type now, I'll try deleting the copy constructor/assignment operator and see.

Reply:
Ah, no. It does get copied. It's used by variables and directives, which both get passed around as a Map, but neither of those is used in the sample for the benchmark.

Reply:
do you think these needs to be copied to the resolvers? Is there any use case where these would be changed?

Anyway, as you can see in my PR, the COW is pretty simple to do. Just use a `shared_ptr` and check if `unique()` before doing any modification (ex: `emplace_back()`). If so, then you create a copy before modifying.